apps/veil/src/pages/portfolio/ui/shield-unshield.tsx

```diff
@@ -0,0 +1,228 @@
+import type { UnifiedAsset } from '@/pages/portfolio/api/use-unified-assets.ts';
+import { Button } from '@penumbra-zone/ui/Button';
+import { Tooltip } from '@penumbra-zone/ui/Tooltip';
+import { Widget } from '@skip-go/widget';
+import { useEffect, useState, useMemo } from 'react';
+import { X } from 'lucide-react';
+import { useRegistry } from '@/shared/api/registry.ts';
+import { getMetadata } from '@penumbra-zone/getters/value-view';
+
+/**
+ * Computes the IBC denom hash for Penumbra-1
+ * If the denom already starts with "ibc/", it's assumed to be already computed.
+ * If the denom starts with "transfer/", it's assumed to be a full trace path.
+ * Otherwise, constructs the trace path using channelId and denom.
+ */
+async function computeIbcDenom(denom: string, channelId: string): Promise<string> {
+  // Check if denom is already an IBC hash
+  if (denom.startsWith('ibc/')) {
+    return denom;
+  }
+  // Check if denom already is a full trace path
+  const ibcTraceStr = denom.startsWith('transfer/') ? denom : `transfer/${channelId}/${denom}`;
```

Comment:
thought: many assets have a `/factory/` substring in them like here: https://github.com/prax-wallet/registry/blob/main/registry/chains/penumbra-1.json#L129

question: should we support it too? Won't the things break if we don't consider this `factory` thing?

Reply:
Good catch - this is one of the sources of complexity for deposits and withdrawals. I will support factory as well, wondering if there are already utilities for this somewhere - either in our repos, or in 3rd party cosmos projects.